go/internal/feast/sqliteonlinestore.go

```diff
@@ -0,0 +1,155 @@
+	in_query := make([]string, len(entityKeys))
+	serialized_entities := make([]interface{}, len(entityKeys))
+	for i := 0; i < len(entityKeys); i++ {
+		serKey, err := serializeEntityKey(entityKeys[i])
+		if err != nil {
+			return nil, err
+		}
+		// TODO: fix this, string conversion is not safe
+		entityNameToEntityIndex[hashSerializedEntityKey(serKey)] = i
+		// for IN clause in read query
+		in_query[i] = "?"
+		serialized_entities[i] = *serKey
+	}
+	featureNamesToIdx := make(map[string]int)
+	for idx, name := range featureNames {
+		featureNamesToIdx[name] = idx
+	}
+
+	for idx := 0; idx < len(entityKeys); idx++ {
+		results[idx] = make([]FeatureData, featureCount)
+	}
+	for _, featureViewName := range featureViewNames {
+		query_string := fmt.Sprintf(`SELECT entity_key, feature_name, value, event_ts
+									FROM %s
+									WHERE entity_key IN (%s)
+									ORDER BY entity_key`, tableId(project, featureViewName), strings.Join(in_query, ","))
+		rows, err := s.db.Query(query_string, serialized_entities...)
+		if err != nil {
+			return nil, err
+		}
+		defer rows.Close()
+		for rows.Next() {
+			var entity_key []byte
+			var feature_name string
+			var valueString []byte
+			var event_ts time.Time
+			var value types.Value
+			err = rows.Scan(&entity_key, &feature_name, &valueString, &event_ts)
+			if err != nil {
+				return nil, errors.New("error could not resolve row in query (entity key, feature name, value, event ts)")
+			}
+			if err := proto.Unmarshal(valueString, &value); err != nil {
+				return nil, errors.New("error converting parsed value to types.Value")
+			}
+			results[entityNameToEntityIndex[hashSerializedEntityKey(&entity_key)]][featureNamesToIdx[feature_name]] = FeatureData{reference: serving.FeatureReferenceV2{FeatureViewName: featureViewName, FeatureName: feature_name},
+				timestamp: *timestamppb.New(event_ts),
+				value:     types.Value{Val: value.Val},
+			}
+		}
+	}
+	return results, nil
+}
+
+// Gets a sqlite connection and sets it to the online store and also returns a pointer to the connection.
+func (s *SqliteOnlineStore) getConnection() (*sql.DB, error) {
+	s.db_mu.Lock()
+	defer s.db_mu.Unlock()
+	if s.db == nil {
+		if s.path == "" {
+			return nil, errors.New("no database path available")
+		}
+		db, err := initializeConnection(s.path)
+		s.db = db
+		if err != nil {
+			return nil, err
+		}
+	}
+	return s.db, nil
+}
+
+// Constructs the table id from the project and table(featureViewName) string.
+func tableId(project string, featureViewName string) string {
+	return fmt.Sprintf("%s_%s", project, featureViewName)
+}
+
+// Creates a connection to the sqlite database and returns the connection.
+func initializeConnection(db_path string) (*sql.DB, error) {
+	db, err := sql.Open("sqlite3", db_path)
+	if err != nil {
+		return nil, err
+	}
+	return db, nil
+}
+
+func hashSerializedEntityKey(serializedEntityKey *[]byte) string {
+	h := sha1.New()
+	h.Write(*serializedEntityKey)
```

Comment:
please check the pointer before dereferencing it

Reply:
Done.